packages/slate-react/src/components/editor.js

```diff
@@ -71,140 +72,112 @@ class Editor extends React.Component {
+    // the updates, then warn the user that they may be doing something wrong.
+    if (this.tmp.resolves > 5 && this.tmp.resolves == this.tmp.updates) {
+      logger.warn(
+        'A Slate <Editor> is re-resolving `props.plugins` or `props.schema` on each update, which leads to poor performance. This is often due to passing in a new `schema` or `plugins` prop with each render by declaring them inline in your render function. Do not do this!'
+      )
+    }
+
+    this.change(change => change)
   }
 
   /**
-   * Queue a `change` object, to be able to flush it later. This is required for
-   * when a change needs to be applied to the value, but because of the React
-   * lifecycle we can't apply that change immediately. So we cache it here and
-   * later can call `this.flushChange()` to flush it.
-   *
-   * @param {Change} change
+   * Get event handlers as passed by onEvent
    */
 
-  queueChange = change => {
-    if (change.operations.size) {
-      debug('queueChange', { change })
-      this.tmp.change = change
+  handlers = EVENT_HANDLERS.reduce((obj, handler) => {
+    obj[handler] = event => {
+      this.onEvent(handler, event)
     }
-  }
+    return obj
+  }, {})
 
   /**
-   * Flush a temporarily stored `change` object, for when a change needed to be
-   * made but couldn't because of React's lifecycle.
+   * Perform a change on the editor, passing `...args` to `change.call`.
+   *
+   * @param {Mixed} ...args
    */
 
-  flushChange = () => {
+  change = (...args) => {
     const { change } = this.tmp
+    const lastOperationSize = change.operations.size
+    change.call(...args)
+    debug('onChange', { change })
+
+    // Do not rerun the change if onChange is run already in associateStackAndValue without more following operations
+    if (change.operations.size > lastOperationSize) {
+      this.stack.run('onChange', change, this)
+    }
 
-    if (change) {
-      debug('flushChange', { change })
-      delete this.tmp.change
-      this.props.onChange(change)
+    // After we update the change with props.onChange; do not re-update in second
+    // cycle if no other operations are added
+    if (change.operations.size === 0 && change.value === this.props.value) {
+      return
     }
+
+    this.updateChange(change)
   }
 
   /**
-   * Perform a change on the editor, passing `...args` to `change.call`.
+   * On change.
    *
-   * @param {Mixed} ...args
+   * @param {Change} change
    */
 
-  change = (...args) => {
-    const change = this.value.change().call(...args)
-    this.onChange(change)
+  onChange = change => {
+    debug('onChange', { change })
+    const { value } = change
+    if (value == this.props.value && change.operations.size === 0) return
+    this.stack.run('onChange', change, this)
+    this.updateChange(change)
+  }
+
+  /*
+   * Update by change, and cache the current stack and value for update
+   * @param {Change} change
+  */
+
+  updateChange = change => {
```

Comment:
Similarly, I think this `updateChange` method should be only called from a single place, and then it wouldn't need to be its own method.